packages/gatsby/src/internal-plugins/functions/gatsby-node.ts

```diff
@@ -43,6 +43,55 @@ interface IGlobPattern {
   globPattern: string
 }
 
+// During development, we lazily compile functions only when they're requested.
+// Here we keep track of which functions have been requested so are "active"
+const activeDevelopmentFunctions = new Set<IFunctionData>()
+let activeEntries = {}
+
+async function _restartWebpack(
+  compiler,
+  config,
+  callback
+): Promise<webpack.Compiler> {
+  return await new Promise(resolve =>
+    compiler.close(async () => {
+      compiler = webpack(config).watch({}, callback)
+      resolve(compiler)
+    })
+  )
+}
+
+// Throttle restarting the compilier to once every 500ms so we
+// don't overwhelm the CI server during tests
+const restartWebpack = _.throttle(_restartWebpack, 500)
+
+async function ensureFunctionIsCompiled(functionObj: IFunctionData): any {
+  // stat the compiled function. If it's there, then return.
+  let compiledFileExists = false
+  try {
+    compiledFileExists = !!(await fs.stat(functionObj.absoluteCompiledFilePath))
```

Comment:
This doesn't tell you that the function was actually done compiling. This only tells you it exists.

Reply:
I assume webpack won't write out a file until it's compiled? Is this ever not the case?

Reply:
Do we clear the functions directory? If not this can give stale functions. Something to fix in a follow up.

Reply:
yeah https://github.com/gatsbyjs/gatsby/pull/31508/files#diff-0943998abf75a4032a0ff6d7584247970234f486cf37a3cefcffb4a4c8ca4cb6R354